auth_saml/models/res_users.py

```diff
@@ -0,0 +1,219 @@
+
+        # TODO replace assert by proper raise... asserts do not execute in
+        # production code...
+        assert len(user_ids) == 1
+        user = user_ids[0]
+
+        # now find if a token for this user/provider already exists
+        token_ids = token_osv.search(
+            [
+                ('saml_provider_id', '=', provider),
+                ('user_id', '=', user.id),
+            ]
+        )
+
+        if token_ids:
+            token_ids.write(
+                {'saml_access_token': saml_response},
+            )
+        else:
+            token_osv.create(
+                {
+                    'saml_access_token': saml_response,
+                    'saml_provider_id': provider,
+                    'user_id': user.id,
+                },
+            )
+
+        return user.login
+
+    @api.model
+    def auth_saml(self, provider, saml_response):
+
+        validation = self._auth_saml_validate(provider, saml_response)
+
+        # required check
+        if not validation.get('user_id'):
+            raise openerp.exceptions.AccessDenied()
+
+        # retrieve and sign in user
+        login = self._auth_saml_signin(provider, validation, saml_response)
+
+        if not login:
+            raise openerp.exceptions.AccessDenied()
+
+        # return user credentials
+        return self.env.cr.dbname, login, saml_response
+
+    # This method is using the old v7 API because it is called BEFORE the login
+    def check_credentials(self, cr, uid, token):
+        """token can be a password if the user has used the normal form...
+        but we are more interested in the case when they are tokens
+        and the interesting code is inside the except clause
+        """
+        token_osv = self.pool.get('auth_saml.token')
+
+        try:
+            baseuser.check_credentials(
+                self,
+                cr,
+                uid,
+                token
+            )
+
+        except (
+            openerp.exceptions.AccessDenied,
+            passlib.exc.PasswordSizeError,
+        ):
+            # since normal auth did not succeed we now try to find if the user
+            # has an active token attached to his uid
+            res = token_osv.search(
+                cr, SUPERUSER_ID,
+                [
+                    ('user_id', '=', uid),
+                    ('saml_access_token', '=', token),
+                ]
+            )
+
+            # if the user is not found we re-raise the AccessDenied
+            if not res:
+                # TODO: maybe raise a defined exception instead of the last
+                # exception that occurred in our execution frame
+                raise
+
+    @api.multi
+    def write(self, vals):
+        """Override to clear out the user's password when setting an SAML user
+        ID (as they can't cohabit).
+        """
+
+        if vals and vals.get('saml_uid'):
+            if not self._allow_saml_and_password():
```

Comment:
Should we avoid to unset password when admin user wants to change its password?

I don't remember in case of ``auth_crypt`` module is installed, when the password is crypted when changing the value from the UI, we may want to del ``password_crypt`` as well?